{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib\n",
    "import scipy\n",
    "import isolearn\n",
    "import itertools\n",
    "import collections\n",
    "from Levenshtein import distance\n",
    "\n",
    "matplotlib.rcParams['figure.dpi'] = 120\n",
    "matplotlib.rcParams['font.sans-serif'] = 'Helvetica'\n",
    "\n",
    "# import dinucleotide shuffling function from deeplift\n",
    "from deeplift.dinuc_shuffle import dinuc_shuffle\n",
    "\n",
    "import re\n",
    "\n",
    "import sys\n",
    "\n",
    "sys.path.append('../../../analysis/paper_figures')\n",
    "sys.path.append('../')\n",
    "from figure_utils import *\n",
    "from analyze_motifs import *\n",
    "\n",
    "from design_utils import TEST_FOLD_IDX,DESIGNED_SEQ_LEN,MODEL_TYPE_DF,CELL_TYPES\n",
    "from design_utils import load_ensemble_model, seq_to_one_hot, longest_repeat, get_paired_editdistances, load_maxmin_ensemble_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load data ###\n",
    "\n",
    "d2_final_df,_,d2_deseq_df = load_motif_data_d2(qthresh_suffix='_qthresh05')\n",
    "# rebase d2_final_df start and stop to 0 indexed\n",
    "d2_final_df['start'] -= 1\n",
    "# d2_final_df['stop'] -= 1 # don't adjust stop; this is the last base that is included in the sequence\n",
    "\n",
    "# sort d2_deseq_df by descending log2FoldChange_H2K_deseq\n",
    "# d2_deseq_df = d2_deseq_df.sort_values(by='log2FoldChange_H2K_deseq', ascending=False)\n",
    "\n",
    "seq_df = pd.DataFrame(columns=['sequence_name','design_type','cell_type','sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[433 905 633 335 667]\n",
      "[ 976 1041  762  156 1099]\n"
     ]
    }
   ],
   "source": [
    "# get indices of n_top_seqs highest log2FoldChange_H2K_deseq enhancers and n_top_seqs lowest log2FoldChange_H2K_deseq enhancers\n",
    "n_top_seqs = 5\n",
    "top_hepg2_inds = np.argsort(d2_deseq_df['log2FoldChange_H2K_deseq'])[::-1][:n_top_seqs].values\n",
    "top_k562_inds = np.argsort(d2_deseq_df['log2FoldChange_H2K_deseq'])[:n_top_seqs].values\n",
    "print(top_hepg2_inds)\n",
    "print(top_k562_inds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "433\n",
      "3\n",
      "905\n",
      "5\n",
      "633\n",
      "3\n",
      "335\n",
      "3\n",
      "667\n",
      "4\n",
      "976\n",
      "2\n",
      "1041\n",
      "2\n",
      "762\n",
      "5\n",
      "156\n",
      "6\n",
      "1099\n",
      "7\n"
     ]
    }
   ],
   "source": [
    "motif_flank_buffer = 1 # how many bps to each side of motif to include in motif\n",
    "seq_len = 145\n",
    "\n",
    "def extract_nonmotif_from_seq(nonmotif_mask):\n",
    "    nonmotif_intervals = []\n",
    "    cur_interval = []\n",
    "    for i in range(len(nonmotif_mask)) :\n",
    "        if nonmotif_mask[i] == 1 :\n",
    "            if len(cur_interval) == 0 :\n",
    "                cur_interval.append(i)\n",
    "        else :\n",
    "            if len(cur_interval) > 0 :\n",
    "                cur_interval.append(i)\n",
    "                nonmotif_intervals.append(cur_interval)\n",
    "                cur_interval = []\n",
    "    return nonmotif_intervals\n",
    "\n",
    "\n",
    "def check_for_RE_sites(seq):\n",
    "\n",
    "    KPNI = \"GGTACC\"\n",
    "    XBAI = \"TCTAGA\"\n",
    "    SFII = \"GGCC.....GGCC\"\n",
    "\n",
    "    KPNI_sites = [m.start() for m in re.finditer(KPNI, seq)]\n",
    "    if len(KPNI_sites) > 0 :\n",
    "        return True\n",
    "    XBAI_sites = [m.start() for m in re.finditer(XBAI, seq)]\n",
    "    if len(XBAI_sites) > 0 :\n",
    "        return True\n",
    "    SFII_sites = [m.start() for m in re.finditer(SFII, seq)]\n",
    "    if len(SFII_sites) > 0 :\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "def generate_shuffled_nonmotif_seqs(seq_idx, cur_cell_type, cur_seq, nonmotif_intervals, n_shuffled_nonmotif_seqs,seq_df):\n",
    "    for i in range(n_shuffled_nonmotif_seqs):\n",
    "        # get shuffled nonmotif sequence - w/o RE sites!\n",
    "        RE_sites = True\n",
    "        while RE_sites:\n",
    "            shuffled_nonmotif_seq = np.array(list(cur_seq))\n",
    "            for nonmotif_interval in nonmotif_intervals :\n",
    "                start,stop = nonmotif_interval\n",
    "                shuffled_nonmotif_seq[start:stop] = np.array(list(dinuc_shuffle(cur_seq[start:stop])))\n",
    "\n",
    "            shuffled_nonmotif_seq = ''.join(shuffled_nonmotif_seq)\n",
    "            RE_sites = check_for_RE_sites(shuffled_nonmotif_seq)\n",
    "\n",
    "        seq_name = f'{seq_idx}_shuffled_nonmotif_{i}'\n",
    "        seq_df = pd.concat([seq_df, pd.DataFrame([[seq_name, 'shuffled_nonmotif', cur_cell_type,shuffled_nonmotif_seq]], columns=['sequence_name','design_type','cell_type','sequence'])])\n",
    "    return seq_df\n",
    "\n",
    "def generate_ablated_motif_seqs(motif_flank_buffer, seq_idx, cur_cell_type, cur_seq, cur_seq_motifs, motif_pairs, seq_df,n_vars):\n",
    "    # do single motif ablations\n",
    "    for motif_idx in range(len(cur_seq_motifs)):\n",
    "        for var_idx in range(n_vars):\n",
    "            RE_sites = True\n",
    "            while RE_sites:\n",
    "                cur_seq_copy = cur_seq\n",
    "                motif_start, motif_end = cur_seq_motifs.iloc[motif_idx]['start'], cur_seq_motifs.iloc[motif_idx]['stop']\n",
    "                # replace motif with shuffled sequence - note here I'm shuffling the whole sequence and excerpting an appropriate length to replace the motif, this is a reasonable bg, shuffling only the motif would probably result in motif-like sequences too often\n",
    "                cur_seq_copy = cur_seq_copy[:motif_start - motif_flank_buffer] + dinuc_shuffle(cur_seq)[motif_start - motif_flank_buffer:motif_end + motif_flank_buffer] + cur_seq_copy[motif_end + motif_flank_buffer:]\n",
    "                RE_sites = check_for_RE_sites(cur_seq_copy)\n",
    "            seq_name = f'{seq_idx}_ablated_{motif_idx}_{var_idx}'\n",
    "            seq_df = pd.concat([seq_df, pd.DataFrame([[seq_name, 'ablated', cur_cell_type,cur_seq_copy]], columns=['sequence_name','design_type','cell_type','sequence'])])\n",
    "\n",
    "    # do motif pair ablations\n",
    "    if len(motif_pairs) > 2 :\n",
    "        for motif_set_idx in range(len(motif_pairs)):\n",
    "            for var_idx in range(n_vars):\n",
    "                RE_sites = True\n",
    "                while RE_sites:\n",
    "                    cur_seq_copy = cur_seq\n",
    "                    motifs = np.array(motif_pairs)[motif_set_idx]\n",
    "                    cur_motif_set = cur_seq_motifs.iloc[motifs]\n",
    "                    for motif_start, motif_end in zip(cur_motif_set['start'].values, cur_motif_set['stop'].values) :\n",
    "                        # replace motif with shuffled sequence - note here I'm shuffling the whole sequence and excerpting an appropriate length to replace the motif, this is a reasonable bg, shuffling only the motif would probably result in motif-like sequences too often\n",
    "                        cur_seq_copy = cur_seq_copy[:motif_start - motif_flank_buffer] + dinuc_shuffle(cur_seq)[motif_start - motif_flank_buffer:motif_end + motif_flank_buffer] + cur_seq_copy[motif_end + motif_flank_buffer:]\n",
    "                    RE_sites = check_for_RE_sites(cur_seq_copy)\n",
    "\n",
    "                powerset_suffix = ''.join([str(i) for i in motif_pairs[motif_set_idx]])\n",
    "                seq_name = f'{seq_idx}_ablated_{powerset_suffix}_{var_idx}'\n",
    "                seq_df = pd.concat([seq_df, pd.DataFrame([[seq_name, 'ablated', cur_cell_type,cur_seq_copy]], columns=['sequence_name','design_type','cell_type','sequence'])])\n",
    "    return seq_df\n",
    "\n",
    "seq_df = pd.DataFrame()\n",
    "\n",
    "for seq_idx in np.concatenate([top_hepg2_inds,top_k562_inds]):\n",
    "    print(seq_idx)\n",
    "    \n",
    "    # seq_idx = d2_deseq_df.index[cur_idx]\n",
    "    cur_cell_type = d2_deseq_df.iloc[seq_idx]['cell_type']\n",
    "\n",
    "    cur_seq = d2_deseq_df.iloc[seq_idx]['enhancer']\n",
    "    cur_seq_motifs = d2_final_df[d2_final_df['sequence_name'] == seq_idx]\n",
    "\n",
    "    # create mask with motif positions\n",
    "    motif_mask = np.zeros(seq_len,dtype=int)\n",
    "    for motif_start, motif_end in zip(cur_seq_motifs['start'].values, cur_seq_motifs['stop'].values) :\n",
    "        motif_mask[motif_start - motif_flank_buffer:motif_end + motif_flank_buffer] = 1\n",
    "\n",
    "    # get nonmotif_mask - this can be used as input to masked SeqProp for optimized/adversarial perturbations!\n",
    "    nonmotif_mask = (1 - motif_mask)\n",
    "\n",
    "    n_motifs = len(cur_seq_motifs)\n",
    "    n_motifs_not_cluster27 = len(cur_seq_motifs[cur_seq_motifs['jaspar_cluster'] != 'cluster_27'])\n",
    "    # get all pairs of motifs\n",
    "    motif_pairs = list(itertools.combinations(range(n_motifs), 2))\n",
    "    # for pair in motif_pairs :\n",
    "    #     print(pair)\n",
    "    print(n_motifs)\n",
    "\n",
    "    # print(f'{seq_idx} {n_motifs} {n_motifs_not_cluster27} {len(motif_pairs)}')\n",
    "\n",
    "    # get all intervals of nonmotif_sequence\n",
    "    nonmotif_intervals = extract_nonmotif_from_seq(nonmotif_mask)\n",
    "\n",
    "    n_shuffled_nonmotif_seqs = 10\n",
    "    n_vars = 3\n",
    "\n",
    "    seq_df = generate_shuffled_nonmotif_seqs(seq_idx, cur_cell_type, cur_seq, nonmotif_intervals, n_shuffled_nonmotif_seqs,seq_df)\n",
    "\n",
    "    seq_df = generate_ablated_motif_seqs(motif_flank_buffer, seq_idx, cur_cell_type, cur_seq, cur_seq_motifs, motif_pairs, seq_df,n_vars)\n",
    "\n",
    "seq_df['og_seq_idx'] = seq_df['sequence_name'].apply(lambda x: int(x.split('_')[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "433 out of 433 sequences have no restriction sites\n"
     ]
    }
   ],
   "source": [
    "def filter_RE_sites(temp_df):\n",
    "\n",
    "    KPNI = \"GGTACC\"\n",
    "    XBAI = \"TCTAGA\"\n",
    "    SFII = \"GGCC.....GGCC\"\n",
    "\n",
    "    temp_df[\"kpnI\"] = temp_df[\"sequence\"].apply(lambda x: KPNI in x)\n",
    "    temp_df[\"xbaI\"] = temp_df[\"sequence\"].apply(lambda x: XBAI in x)\n",
    "    temp_df[\"sfiI\"] = temp_df[\"sequence\"].apply(lambda x: re.search(SFII, x))\n",
    "        # convert None to 0, False to 0, True to 1\n",
    "    temp_df[\"sfiI\"] = temp_df[\"sfiI\"].apply(lambda x: 0 if x is None else 1)\n",
    "    temp_df[\"kpnI\"] = temp_df[\"kpnI\"].apply(lambda x: 1 if x is True else 0)\n",
    "    temp_df[\"xbaI\"] = temp_df[\"xbaI\"].apply(lambda x: 1 if x is True else 0)\n",
    "\n",
    "        # extract only rows with no restriction sites\n",
    "    temp_df = temp_df.loc[(temp_df['kpnI']==0) & (temp_df['xbaI']==0) & (temp_df['sfiI']==0)]\n",
    "    return temp_df\n",
    "\n",
    "n_total_seqs = seq_df.shape[0]\n",
    "temp_df = filter_RE_sites(seq_df)\n",
    "print(f'{temp_df.shape[0]} out of {n_total_seqs} sequences have no restriction sites')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1099    84\n",
       "156     63\n",
       "905     45\n",
       "762     45\n",
       "667     30\n",
       "433     18\n",
       "633     18\n",
       "335     18\n",
       "976      6\n",
       "1041     6\n",
       "Name: og_seq_idx, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_df[(seq_df['design_type']=='ablated')]['og_seq_idx'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14/14 [==============================] - 6s 373ms/step\n",
      "14/14 [==============================] - 34s 2s/step\n"
     ]
    }
   ],
   "source": [
    "## don't need to run this again ###\n",
    "\n",
    "\n",
    "model_types = ['d1_finetuned','dhs64_finetuned']\n",
    "design_seq_len = 145\n",
    "x_tot = np.array([seq_to_one_hot(seq) for seq in seq_df['sequence'].values])\n",
    "\n",
    "for model_type in model_types:\n",
    "    model_basename = MODEL_TYPE_DF.loc[model_type,'model_basename']\n",
    "    max_seq_len = MODEL_TYPE_DF.loc[model_type,'input_len']\n",
    "    n_models_to_ensemble = MODEL_TYPE_DF.loc[model_type,'n_ensemble']\n",
    "\n",
    "    model_dir = f'../../aws/for_aws/sequencing_analysis/retraining/retrained_models/cf10/test_fold_0/{model_type}'\n",
    "    if model_type == 'd2_dhs':\n",
    "        model_dir = f'../../aws/for_aws/sequencing_analysis/retraining/retrained_models/cf10/test_fold_0/{model_type}/hp_0'\n",
    "    # if model_type in ['d1_finetuned']:\n",
    "    #     continue\n",
    "\n",
    "    K.clear_session()\n",
    "\n",
    "    model_ensemble = load_ensemble_model(model_dir,model_basename,range(1,1+n_models_to_ensemble),\n",
    "                                        design_seq_len,max_seq_len=max_seq_len)\n",
    "    \n",
    "\n",
    "    y_pred = model_ensemble.predict(x_tot)\n",
    "\n",
    "    # add columns for the predictions to seq_df with the model type as a suffix\n",
    "    seq_df[f'log2(HEPG2)_pred_{model_type}'] = y_pred[:,0]\n",
    "    seq_df[f'log2(K562)_pred_{model_type}'] = y_pred[:,1]\n",
    "    seq_df[f'log2(H2K)_pred_{model_type}'] = y_pred[:,0]-y_pred[:,1]\n",
    "\n",
    "# write seq_df to csv\n",
    "design_dir = '../D3_design_scripts/designed_seqs/'\n",
    "seq_df.to_csv(f'{design_dir}/shuffled_and_ablated_top_d2_enhancers.csv',index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
